docs/coding-guidelines/project-guidelines.md

```diff
@@ -80,37 +76,74 @@ When building an individual project the `BuildTargetFramework` and `TargetOS` wi
+<ItemGroup Condition="'$(TargetFrameworkIdentifier)' == '.NETFramework'">...</ItemGroup>
+```
+2. Use equality checks if you want to condition on specific runtime agnostic target frameworks (i.e. without the `-windows` suffix).
 Example:
 ```
 <PropertyGroup>
-  <TargetFrameworks>netstandard2.0;netstandard2.1</TargetFrameworks>
+  <TargetFrameworks>$(NetCoreAppCurrent);netstandard2.0;$(TargetFrameworkMinimum)</TargetFrameworks>
 </PropertyGroup>
+<ItemGroup Condition="'$(TargetFramework)' == '$(NetCoreAppCurrent)'">...</ItemGroup>
 <ItemGroup Condition="'$(TargetFramework)' == 'netstandard2.0'">...</ItemGroup>
+<ItemGroup Condition="'$(TargetFramework)' == '$(TargetFrameworkMinimum)'">...</ItemGroup>
+```
+3. Use the `TargetPlatformIdentifier` property to condition on a .NETCoreApp platform specific target framework. Note that .NETStandard and .NETFramework target frameworks can't target a specific platform anymore.
+Example:
+```
+<PropertyGroup>
+  <TargetFrameworks>$(NetCoreAppCurrent)-windows;$(NetCoreAppCurrent)-OSX;$(NetCoreAppCurrent)</TargetFrameworks>
+</PropertyGroup>
+<ItemGroup Condition="'$(TargetPlatformIdentifier)' == 'windows'">...</ItemGroup>
+<ItemGroup Condition="'$(TargetPlatformIdentifier)' == 'OSX'">...</ItemGroup>
 ```
-2. Use a StartsWith when you want to test for multiple .NETStandard or .NETFramework versions.
+Important: In contrast to the old `Targets*` checks, `TargetPlatformIdentifier` conditions apply to a single tfm only, inheritance between target frameworks can't be expressed. See the example below for Unix:
+```
+<PropertyGroup>
+  <TargetFrameworks>$(NetCoreAppCurrent)-Unix;$(NetCoreAppCurrent)-Linux;$(NetCoreAppCurrent)-android;$(NetCoreAppCurrent)-windows</TargetFrameworks>
+</PropertyGroup>
+<ItemGroup Condition="'$(TargetPlatformIdentifier)' == 'Unix' or '$(TargetPlatformIdentifier)' == 'Linux' or '$(TargetPlatformIdentifier)' == 'android'">...</ItemGroup>
+<!-- Negations make such conditions easier to write and read. -->
+<ItemGroup Condition="'$(TargetPlatformIdentifier)' != 'windows'">...</ItemGroup>
+```
+
+4. Use the `TargetPlatformIdentifier` property in combination with the `TargetFrameworkIdentifier` property to exclude
+.NETFramework and .NETStandard target frameworks as their platform identifier is `windows` by default.
 Example:
 ```
 <PropertyGroup>
-  <TargetFrameworks>netstandard2.0;netstandard2.1</TargetFrameworks>
+  <TargetFrameworks>$(NetCoreAppCurrent)-windows;$(NetCoreAppCurrent);netstandard2.0;net462</TargetFrameworks>
 </PropertyGroup>
-<ItemGroup Condition="$(TargetFramework.StartsWith('netstandard'))>...</ItemGroup>
+<ItemGroup Condition="'$(TargetFrameworkIdentifier)' == '.NETCoreApp' and '$(TargetPlatformIdentifier)' == 'windows'">...</ItemGroup>
 ```
-3. Use a StartsWith if the TargetFramework is overloaded with the OS portion.
+5. Set the `TargetPlatformIdentifier` property in the project to be able to condition on it in properties in the project file.
```

Comment:
I was thinking about this and wondering if we could get away with calculating this in a props if we guarantee the following:
1. Whenever a framework with a TargetPlatformIdentifer exists, there must be a framework without a TargetPlatformIdentifier.  We need this for P2Ps to work well, and most of the times it's good practice to ensure we have a default implementation.  We could enforce this with a target.
2. 1 ensures that any project with a TargetPlatformIdentifier will have more than one TargetFramework, and must have an outer build.
3. Therefore, we can assume that whenever a TargetPlatformIdentifier is present in the TargetFramework it will be passed as a global property, and we can compute the TargetPlatformIdentifier in props.

Reply:
We might be able to even get away without forcing 1, if we just caught the one case where it doesn't work, and forced that case to define it in the project.

We can observe the value of TargetPlatformIdentifier at the start of our Directory.Build.targets evaluation, so we can see what its value is for any of our projects and capture that early on in property evaluation.  We can then see what the value is after the SDK computes it later.  Whenever they mismatch we can raise an error and tell the developer to add a TargetPlatformEvaluation into the project.

Reply:
Nevermind, now I remember why we cannot garuntee this --> design time build.  It never passes TargetFramework values as a global property.  It still sets them with "SetProperty" which effectively appends to the first PropertyGroup.  That's a real bummer.

If we were to get that fixed I think we could clean up this boilerplate.

Reply:
I was trying to avoid any TargetFramework reads or write before the project file itself is evaluated even if the TargetFramework property would be passed in as a global property mostly because other components like the SDK don't do that either. Even though it's boilerplate, it's WYSIWYG so I'm ok with it.